src/System.Text.Primitives/System/Text/Encoding/Utf8/Utf8Encoder.cs

```diff
@@ -288,55 +298,121 @@ internal static bool TryComputeEncodedBytes(ReadOnlySpan<uint> utf32, out int by
         /// <param name="charactersConsumed">On exit, contains the number of characters that were consumed from the UTF-16 span.</param>
         /// <param name="bytesWritten">An output parameter to store the number of bytes written to <paramref name="utf8"/></param>
         /// <returns>True if the input buffer was fully encoded into the output buffer, otherwise false.</returns>
-        public static bool TryEncode(ReadOnlySpan<char> utf16, Span<byte> utf8, out int charactersConsumed, out int bytesWritten)
+        public static unsafe bool TryEncode(ReadOnlySpan<char> utf16, Span<byte> utf8, out int charactersConsumed, out int bytesWritten)
         {
-            var availableCharacters = utf16.Length;
-            var bufferLength = utf8.Length;
-
-            charactersConsumed = 0;
-            bytesWritten = 0;
-
-            for (int i = 0; i < availableCharacters; i++)
+            fixed (char* pUtf16 = &utf16.DangerousGetPinnableReference())
             {
-                var ch = utf16[i];
-
-                if ((ushort)ch <= 0x7f) // this if block just optimizes for ascii
+                fixed (byte* pUtf8 = &utf8.DangerousGetPinnableReference())
                 {
-                    if (bufferLength - bytesWritten < 1)
-                        return false;
+                    var outputBytesAvailable = utf8.Length;
+                    var inputCharacters = utf16.Length;
 
-                    utf8[bytesWritten++] = (byte)ch;
-                    charactersConsumed++;
-                }
-                else
-                {
-                    uint codePoint;
-                    int consumed;
+                    var bytesWrittenTemp = 0;
+                    var i = 0;
+                    var codePoint = '\0';
 
-                    if (!char.IsSurrogate(ch))
+                    ascii:
+                    for (; i < inputCharacters; i++)
                     {
-                        codePoint = (uint)ch;
-                        consumed = 1;
+                        codePoint = *(pUtf16 + i);
+
+                        if (codePoint > Utf8OneByteLastCodePoint)
+                        {
+                            goto non_ascii;
+                        }
+                        if (bytesWrittenTemp > outputBytesAvailable - 1)
+                        {
+                            goto need_more;
+                        }
```

Comment:
I agree and I was playing around in this area to get some of the insights from the CoreCLR implementation but I am having trouble working around the output bytes available check. For me to accurately consume characters and set bytes written, I need to have this if (bytesWrittenTemp > outputBytesAvailabe - X) check everywhere because I can't be sure if the character is 1 byte or up to 4.

For example, in this loop unrolling in CoreCLR, how exactly do we check and return when the output buffer is full?
```C#
                    while (pSrc < pStop) {
                        ch = *(int*)pSrc;
                        int chc = *(int*)(pSrc+2);
                        if (((ch | chc) & unchecked((int)0xFF80FF80)) != 0) {
                            goto LongCodeWithMask;
                        }

                        // Unfortunately, this is endianess sensitive
#if BIGENDIAN
                        *pTarget = (byte)(ch>>16);
                        *(pTarget+1) = (byte)ch;
                        pSrc += 4;
                        *(pTarget+2) = (byte)(chc>>16);
                        *(pTarget+3) = (byte)chc;
                        pTarget += 4;
#else // BIGENDIAN
                        *pTarget = (byte)ch;
                        *(pTarget+1) = (byte)(ch>>16);
                        pSrc += 4;
                        *(pTarget+2) = (byte)chc;
                        *(pTarget+3) = (byte)(chc>>16);
                        pTarget += 4;
#endif // BIGENDIAN
                    }
```